apps/dashboard/pkg/migration/conversion/conversion.go

```diff
@@ -9,76 +12,185 @@ import (
 	dashv1 "github.com/grafana/grafana/apps/dashboard/pkg/apis/dashboard/v1beta1"
 	dashv2alpha1 "github.com/grafana/grafana/apps/dashboard/pkg/apis/dashboard/v2alpha1"
 	dashv2beta1 "github.com/grafana/grafana/apps/dashboard/pkg/apis/dashboard/v2beta1"
+	"github.com/grafana/grafana/apps/dashboard/pkg/migration"
+	"github.com/grafana/grafana/apps/dashboard/pkg/migration/schemaversion"
 )
 
 var logger = logging.DefaultLogger.With("logger", "dashboard.conversion")
 
+// withConversionMetrics wraps a conversion function with metrics and logging for the overall conversion process
+func withConversionMetrics(sourceVersionAPI, targetVersionAPI string, conversionFunc func(a, b interface{}, scope conversion.Scope) error) func(a, b interface{}, scope conversion.Scope) error {
+	return func(a, b interface{}, scope conversion.Scope) error {
+		// Extract dashboard UID and schema version from source
+		var dashboardUID string
+		var sourceSchemaVersion interface{} = "unknown"
+		var targetSchemaVersion interface{} = "unknown"
+
+		// Try to extract UID and schema version from source dashboard
+		switch source := a.(type) {
+		case *dashv0.Dashboard:
+			dashboardUID = string(source.UID)
+			if source.Spec.Object != nil {
+				sourceSchemaVersion = source.Spec.Object["schemaVersion"]
+			}
+		case *dashv1.Dashboard:
+			dashboardUID = string(source.UID)
+			if source.Spec.Object != nil {
+				sourceSchemaVersion = source.Spec.Object["schemaVersion"]
+			}
+		case *dashv2alpha1.Dashboard:
+			dashboardUID = string(source.UID)
+			sourceSchemaVersion = "v2alpha1" // V2 doesn't use numeric schema versions
+		case *dashv2beta1.Dashboard:
+			dashboardUID = string(source.UID)
+			sourceSchemaVersion = "v2beta1" // V2 doesn't use numeric schema versions
+		}
+
+		// Determine target schema version based on target type
+		switch b.(type) {
+		case *dashv0.Dashboard:
+			targetSchemaVersion = sourceSchemaVersion // V0 keeps source schema version
+		case *dashv1.Dashboard:
+			targetSchemaVersion = schemaversion.LATEST_VERSION // V1 migrates to latest
+		case *dashv2alpha1.Dashboard:
+			targetSchemaVersion = "v2alpha1"
+		case *dashv2beta1.Dashboard:
+			targetSchemaVersion = "v2beta1"
```

Comment:
WDYT about not defining `sourceSchemaVersion` and `targetSchemaVersion` for conversions different from v0 -> v1? So we avoid errors using schemaVersions instead of apiVersions for monitoring conversions.